### PR TITLE
api: Check condition `scheduled` for `v.Ready`

### PR DESCRIPTION
**Pull request by shuo-wu**
longhorn/longhorn#1360

### CODE REVIEW

File: api/model.go [hunk 1]
```diff
@@ -745,9 +745,14 @@ func toVolumeResource(v *longhorn.Volume, ves []*longhorn.Engine, vrs []*longhor
...
 
-	// If the volume is auto attached or faulted, it's not ready.
+	// The volume is not ready for workloads if:
+	//   1. It's auto attached.
+	//   2. It fails to schedule replicas before the attachment.
+	//   3. It's faulted.
 	ready := true
+	scheduledCondition := types.GetCondition(v.Status.Conditions, types.VolumeConditionTypeScheduled)
 	if (v.Spec.NodeID == "" && v.Status.State != types.VolumeStateDetached) ||
+		(v.Status.State == types.VolumeStateDetached && scheduledCondition.Status != types.ConditionStatusTrue) ||
```
Comment (yasker) [inline api/model.go:755]: What if it's an existing volume, lost one replica and unable to rebuild, then detached? Will this make it unable to be attached again?

Comment (shuo-wu) [reply api/model.go:755]: No. 

For example, if we disable soft anti-affinity, disable a node scheduling and crash the related replica for a running volume, the new replica won't be scheduled and fail the rebuild. But it will be cleaned up during the detachment. After the detachment, there are only healthy replicas in the volume. Hence we can attach/expand the volume later.

To prevent the volume is wrongly marked as NotReady in the above case, I add `v.Status.State == types.VolumeStateDetached` besides `scheduledCondition.Status != types.ConditionStatusTrue`

I was thinking you had considered this issue, otherwise the hard anti-affinity was a dangerous setting.

Comment (yasker) [reply api/model.go:755]: That's the exactly case I was thinking about when I commented. Didn't realize we will clean up scheduledCondition when detached. Then this is fine.

If we don't have a test case for this, we need to add one.

Comment (yasker) [reply api/model.go:755]: Now I saw `before attachment` in the comment might be referring to this? It's not obvious. We can say `during the volume creation, in which case scheduling failure will happen when the volume is detached. In other cases, scheduling failure only happens when the volume is attached`.

Comment (shuo-wu) [reply api/model.go:755]: I plan to add Test 2 into the integration test, is that OK?
longhorn/longhorn#1373 (comment)

Comment (shuo-wu) [reply api/model.go:755]: YES. I will update the comment.

Comment (yasker) [reply api/model.go:755]: @shuo-wu longhorn/longhorn#1373 concerns about volume expansion. I was referring to normal volume operation - detach when degraded, make sure we can attach and use it again with pod (since v.Ready won't be check by the UI but the CSI driver)

Comment (shuo-wu) [reply api/model.go:755]: In this case, the volume should be ready for the workload after the detachment. Users can reuse the volume via UI or the CSI plugin. 

Do we need to add a test for #1373? If YES, there may be duplicate between the tests of these 2 issues. We can add some extra steps for the expansion part after testing this issue.

Comment (yasker) [reply api/model.go:755]: #1373 only concerns about volume expansion. The case here is more common. And yes, we should test both.
